### PR TITLE
Update metrics API

### PR DESCRIPTION
**Pull request by fbogsany**
Update the metrics API to reflect the latest version:
https://github.com/open-telemetry/opentelemetry-specification/blob/a131ae1ccabee8cc8f3b8cf109b81be378006fea/specification/api-metrics.md
and
https://github.com/open-telemetry/opentelemetry-specification/blob/5802f6e33644c5996c2fb46ef9f444f81ac2b158/specification/api-metrics-user.md

### CODE REVIEW

File: api/lib/opentelemetry/metrics/instruments.rb [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    # produced -- mathematical, statistical summaries of certain observable
+    # behavior in the program. "Instruments" are the devices used by the
+    # program to record observations about their behavior. Therefore, we use
+    # "metric instrument" to refer to a program object, allocated through the
+    # API, used for recording metrics. There are three distinct instruments in
+    # the Metrics API, commonly known as Counters, Gauges, and Measures.
+    module Instruments
+      # A float gauge instrument.
+      class FloatGauge
+        def set(value, label_set_or_label = nil, *labels); end
```
Comment (mwear) [inline api/lib/opentelemetry/metrics/instruments.rb]: Are the `label_or_label_set` and `*labels` arguments here for overloading `set` to work with the [label set calling convention](https://github.com/open-telemetry/opentelemetry-specification/blob/5802f6e33644c5996c2fb46ef9f444f81ac2b158/specification/api-metrics-user.md#label-set-calling-convention) and the [option to bypass meter.Lables()`?

Comment (fbogsany) [reply api/lib/opentelemetry/metrics/instruments.rb]: Yep. I'm open to splitting the methods for each calling convention.

Comment (mwear) [reply api/lib/opentelemetry/metrics/instruments.rb]: I suppose we can see how the implementation looks. We can leave them as is for now and decide to change them if it makes sense.

**Comment by fbogsany** [comment]
Remaining work here is to add tests. I'll get that done tonight/tomorrow morning, and aim to get this merged by tomorrow's SIG meeting.

File: api/lib/opentelemetry/metrics/instruments.rb (hunk 1 above)
Comment (fbogsany) [reply api/lib/opentelemetry/metrics/instruments.rb]: It turned out this was wrong. I've fixed in recent commits.

**Comment by fbogsany** [comment]
Tests added. This is ready for review again.

**Pull request merged**